src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutVertexNode.cs

```diff
@@ -0,0 +1,405 @@
+                _parametersSig[i] = factory.NativeLayoutTypeSignatureVertex(method.Signature[i]);
+        }
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            DependencyList dependencies = new DependencyList();
+
+            dependencies.Add(new DependencyListEntry(_returnTypeSig, "NativeLayoutMethodSignatureVertexNode return type signature"));
+            foreach (var arg in _parametersSig)
+                dependencies.Add(new DependencyListEntry(arg, "NativeLayoutMethodSignatureVertexNode parameter signature"));
+
+            return dependencies;
+        }
+
+        public override Vertex WriteVertex(NodeFactory factory, NativeWriter writer)
+        {
+            MethodCallingConvention methodCallingConvention = default(MethodCallingConvention);
+
+            if (_method.Signature.GenericParameterCount > 0)
+                methodCallingConvention |= MethodCallingConvention.Generic;
+            if (_method.Signature.IsStatic)
+                methodCallingConvention |= MethodCallingConvention.Static;
+
+            Debug.Assert(_method.Signature.Length == _parametersSig.Length);
+
+            Vertex returnType = _returnTypeSig.WriteVertex(factory, writer);
+            Vertex[] parameters = new Vertex[_parametersSig.Length];
+            for (int i = 0; i < _parametersSig.Length; i++)
+                parameters[i] = _parametersSig[i].WriteVertex(factory, writer);
+
+            Vertex signature = writer.GetMethodSigSignature((uint)methodCallingConvention, (uint)_method.Signature.GenericParameterCount, returnType, parameters);
+            SavedVertex = factory.MetadataManager.NativeLayoutInfo.SignaturesSection.Place(signature);
+
+            return SavedVertex;
+        }
+    }
+
+    internal sealed class NativeLayoutMethodNameAndSignatureVertexNode : NativeLayoutVertexNode
+    {
+        private MethodDesc _method;
+        private NativeLayoutMethodSignatureVertexNode _methodSig;
+
+        public NativeLayoutMethodNameAndSignatureVertexNode(NodeFactory factory, MethodDesc method)
+        {
+            _method = method;
+            _methodSig = factory.NativeLayoutMethodSignatureVertex(method);
+        }
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            return new DependencyListEntry[] { new DependencyListEntry(_methodSig, "NativeLayoutMethodNameAndSignatureVertexNode signature vertex") };
+        }
+        public override Vertex WriteVertex(NodeFactory factory, NativeWriter writer)
+        {
+            Vertex methodSig = _methodSig.WriteVertex(factory, writer);
+            SavedVertex = writer.GetMethodNameAndSigSignature(_method.Name, methodSig);
+            return SavedVertex;
+        }
+    }
+
+    internal abstract class NativeLayoutTypeSignatureVertexNode : NativeLayoutVertexNode
+    {
+        protected TypeDesc _type;
```

Comment:
The ideal pattern would be for this class to have a protected constructor that takes a `TypeDesc`, and to make the field readonly.

That way it's not possible to forget to set `_type` in child nodes constructors and it's easy to reason about thread safety.